src/index/CompressedBlockPrefiltering.cpp

```diff
@@ -0,0 +1,170 @@
+                                      upperIndex(input, evaluationColumn));
+  } else {
+    // CompOP::NE
+    // Get the indices that define the boundaries w.r.t. blocks potentially
+    // containing non-equal values.
+    auto firstBound = lowerIndex(input, evaluationColumn);
+    auto secondBound = upperIndex(input, evaluationColumn);
+    if (firstBound > secondBound) {
+      return input;  // all blocks are relevant
+    }
+    // Extract the values from the respective ranges defined by the bounding
+    // indices from vector input.
+    // [input.begin(), firstBound) and [secondBound, input.end()]
+    std::vector<BlockMetadata> vecNe;
+    vecNe.reserve(std::distance(input.begin(), firstBound) +
+                  std::distance(secondBound, input.end()));
+    vecNe.insert(vecNe.end(), input.begin(), firstBound);
+    vecNe.insert(vecNe.end(), secondBound, input.end());
+    return vecNe;
+  }
+};
+
+// SECTION LOGICAL OPERATIONS
+//______________________________________________________________________________
+template <LogicalOperators Operation>
+std::vector<BlockMetadata> LogicalExpressions<Operation>::evaluateOr(
+    std::vector<BlockMetadata>& input, size_t evaluationColumn)
+    requires(Operation == LogicalOperators::OR) {
+  auto relevantBlocksChild1 = child1_->evaluate(input, evaluationColumn);
+  auto relevantBlocksChild2 = child2_->evaluate(input, evaluationColumn);
+  // The result vector, reserve space for the logical upper bound (num blocks)
+  // |relevantBlocksChild1| + |relevantBlocksChild2| regarding the OR
+  // operation (union of blocks)
+  std::vector<BlockMetadata> unionedRelevantBlocks;
+  unionedRelevantBlocks.reserve(relevantBlocksChild1.size() +
+                                relevantBlocksChild2.size());
+  // lambda function that implements the less-than comparison required for
+  // std::set_union
+  auto lessThan = [evaluationColumn](const BlockMetadata& block1,
+                                     const BlockMetadata& block2) {
+    return getIdFromColumnIndex(block1.lastTriple_, evaluationColumn) <
+           getIdFromColumnIndex(block2.lastTriple_, evaluationColumn);
+  };
+  // Given that we have vectors with sorted (BlockMedata) values, we can
+  // use std::ranges::set_union, thus the complexity is O(n + m).
+  std::ranges::set_union(relevantBlocksChild1, relevantBlocksChild2,
+                         std::back_inserter(unionedRelevantBlocks), lessThan);
+  return unionedRelevantBlocks;
+};
+
+//______________________________________________________________________________
+template <LogicalOperators Operation>
+std::vector<BlockMetadata> LogicalExpressions<Operation>::evaluate(
+    std::vector<BlockMetadata>& input, size_t evaluationColumn) {
+  if constexpr (Operation == LogicalOperators::AND) {
+    auto resultChild1 = child1_->evaluate(input, evaluationColumn);
+    return child2_->evaluate(resultChild1, evaluationColumn);
+  } else {
+    return evaluateOr(input, evaluationColumn);
```

Comment:
1. You should add a `static_assert(Operation == OR)` here.

Especially since your enum contains a `NOT` expression which you currently don't implement/instantiate. 
It looks like something that you could easily add, or did I miss something?

Reply:
Regarding the `NOT` expression, the implementation is not as simple as just taking the set difference here since blocks can logically overlap (e.g. a block can contain `==`  and ``!=` values w.r.t. a ValueID). But i have an idea on how to solve it, i can probably switch the expressions by using De Morgan's law.

Reply:
I assume that this approach is more efficient than iterating over the blocks and check if they overlap, and forming the "set-difference" that way.